### PR TITLE
PLAD-54  - As an employer, I want to get a recommendation list of students based on how strong the student skills match the placement’s requirements

### PR DESCRIPTION
**Pull request by Fabs97**

### CODE REVIEW

File: backend/DAO/placementDAO.js [hunk 1]
```diff
@@ -134,7 +134,19 @@ module.exports = {
...
+        return database('placements')
+            .select('id',
+            'position',
+            'working_hours',
+            'start_period',
+            'end_period',
+            'salary',
+            'institution',
+            'description_role',
+            'major');
```
Comment (aidadenisa) [inline backend/DAO/placementDAO.js:147]: we don't have majors anymore in the placements table; we have the table "placement_has_major"

File: backend/DAO/studentDAO.js [hunk 2]
```diff
@@ -58,15 +58,22 @@ module.exports = {
...
+        let skillsNumber = skillsId.length;
+        let result = await database('student_has_skills')
+            .select('student_id')
+            .whereIn('skill_id', skillsId)
+            
+            .groupBy('student_id')
+                 
                 //THIS SHOULD BE REPLACED WITH THE 50% CONDITION
-    //         .having(database.raw('count(shs.skill_id) > 2'))
-    // }
```
Comment (aidadenisa) [inline backend/DAO/studentDAO.js:70]: lets remove comments like this one if they are not useful anymore

File: backend/DAO/studentDAO.js [hunk 3]
```diff
@@ -58,15 +58,22 @@ module.exports = {
             resolve(studentToSkills);
         });
     },
-    // getStudentsBySkills: (skillsId) => {
-    //     return database('student')
-    //         .select(['student.id', 'student.name'])
-    //         .leftJoin('student_has_skills AS shs', 'shs.student_id', 'student.id')
-    //         .whereIn('shs.skill_id', skillsId)
-    //         .groupBy('student.id')
```
Comment (aidadenisa) [inline backend/DAO/studentDAO.js:66]: let's remove comments like this one if they are not useful anymore

File: backend/services/recommendationService.js [hunk 4]
```diff
@@ -1,14 +1,19 @@
...
+const studentService = require('./studentService');
 
 module.exports = {
 
     //for employers
-    // getStudentRecommendationsForPlacement: async (placementID) => {   
-    //     let placementSkills = await placementService.getPlacementSkills(placementID);
-    //     let students = await studentService.getStudentsBySkills(placementSkills);
-    //     return students;
-    // },
```
Comment (aidadenisa) [inline backend/services/recommendationService.js:11]: remove code that we don't use anymore

File: backend/services/studentService.js [hunk 5]
```diff
@@ -28,8 +29,21 @@ module.exports = {
         }
         return studentDAO.setStudentSkills(studentId, skills)    
     },
-    // getStudentsBySkills: (skills) => {
-    //     let skillIds = skills.map(skill => skill.skill_id);
-    //     return studentDAO.getStudentsBySkills(skillIds);
-    // },
```
Comment (aidadenisa) [inline backend/services/studentService.js:34]: remove this

File: backend/services/recommendationService.js [hunk 6]
```diff
@@ -1,14 +1,19 @@
...
+    getStudentRecommendationsForPlacement: async (placementID) => {   
+        let placementSkills = await placementsService.getPlacementSkills(placementID);
+        let studentIds = await studentService.getStudentsBySkills(placementSkills);
+        let students =  [];
+        for(let i = 0; i < studentIds.length; i++) {
+            students[i]= await studentService.getStudentProfile(studentIds[i].student_id);
```
Comment (aidadenisa) [inline backend/services/recommendationService.js:13]: This is very bad!! It works in JavaScript but it's a wrong way to add to an array! Please use: 
`let student = await studentService[...];
students.push(student)`

File: backend/services/studentService.js [hunk 7]
```diff
@@ -28,8 +29,21 @@ module.exports = {
...
+
+    getStudentsBySkills: async (skills) => {
+        let skillIds = skills.map(skill => skill.skill_id);
+        return await studentDAO.getStudentsBySkills(skillIds);
+    },
+
+    getStudentProfile: async (id) => {
+        let profile = await studentDAO.getStudentById(id);
+        profile = profile[0];
```
Comment (aidadenisa) [inline backend/services/studentService.js:40]: let's put this line where we get the first element of the resulted query in the DAO method, in "getStudentById". The service should not care that Knex returns us all results in an array, we should send to this directly the single result that we are interested in. When you add this in "getStudentById" you have to make sure you make the method async as well

File: backend/services/studentService.js [hunk 8]
```diff
@@ -28,8 +29,21 @@ module.exports = {
...
+        let skillIds = await skillsDAO.getStudentSkillsById(profile.id);
```
Comment (aidadenisa) [inline backend/services/studentService.js:41]: The student service should not directly access DAO of skills. It should request whatever it wants from skillService. The services communicate with each other.
Besides this, we are making 2 queries here to the same table to get a piece of data that could have been very easily taken in only 1 query. 

So we should have something like: 
`profile.skills = await skillsService.getStudentSkills(profile.id)`

and in the getStudentSkills we should call a method in DAO that gets all the student skills by the student id.

and that method should make a join query on the db between "student_has_skills" and "skills" table, and return the skills in which the student_id is equal with the studentID param

**Review by AliceCasali — APPROVED** [review]
Everything good!

File: backend/services/studentService.js [hunk 9]
```diff
@@ -28,8 +29,21 @@ module.exports = {
...
+        let skills = [];
+        for(let i = 0; i < skillIds.length; i++) {
+            skills[i] = await skillsDAO.getSkillById(skillIds[i].skill_id);
```
Comment (aidadenisa) [inline backend/services/studentService.js:44]: These rows can be deleted if we implement what i mentioned on top

**Review by aidadenisa — CHANGES_REQUESTED** [review]

File: backend/DAO/placementDAO.js (hunk 1 above)
Comment (AnnaBergamasco) [reply backend/DAO/placementDAO.js]: I eliminated those lines

File: backend/DAO/studentDAO.js (hunk 2 above)
Comment (AnnaBergamasco) [reply backend/DAO/studentDAO.js:70]: I eliminated this line

File: backend/DAO/studentDAO.js (hunk 3 above)
Comment (AnnaBergamasco) [reply backend/DAO/studentDAO.js:66]: Those lines where deleted

File: backend/services/recommendationService.js (hunk 4 above)
Comment (AnnaBergamasco) [reply backend/services/recommendationService.js:11]: those lines where deleted

File: backend/services/studentService.js (hunk 5 above)
Comment (AnnaBergamasco) [reply backend/services/studentService.js:34]: those lines where deleted

File: backend/services/recommendationService.js (hunk 6 above)
Comment (AnnaBergamasco) [reply backend/services/recommendationService.js]: I fixed this here and elswhere I used it

File: backend/services/studentService.js (hunk 7 above)
Comment (AnnaBergamasco) [reply backend/services/studentService.js]: I changed it to the version you suggested and made it async

File: backend/services/studentService.js (hunk 8 above)
Comment (AnnaBergamasco) [reply backend/services/studentService.js]: I changed this to the version you suggested, and also changed the code elsewere I did the same mistake as here

File: backend/services/studentService.js (hunk 9 above)
Comment (AnnaBergamasco) [reply backend/services/studentService.js]: I changed it while fixing the issue above

**Review by aidadenisa — APPROVED** [review]

**Pull request merged**